Sources/SourceKitLSP/Swift/MacroExpansionReferenceDocumentURLData.swift

```diff
@@ -0,0 +1,133 @@
+
+import Foundation
+import LanguageServerProtocol
+
+/// Represents url of macro expansion reference document as follows:
+/// `sourcekit-lsp://swift-macro-expansion/LaCb-LcCd.swift?sourceFileURL=&fromLine=&fromColumn=&toLine=&toColumn=&bufferName=`
+///
+/// Here,
+///  - `LaCb-LcCd.swift`, the `displayName`, represents where the macro will expand to or
+/// replace in the source file (i.e. `macroExpansionEditRange`)
+///  - `sourceFileURL` denoting the URL of the source file
+///  - `fromLine`, `fromColumn`, `toLine`, `toColumn` represents the cursor's `selectionRange`
+///  - `bufferName` denotes the buffer name of the specific macro expansion edit
+package struct MacroExpansionReferenceDocumentURLData {
+  package var sourceFileURL: URL
+  package var selectionRange: Range<Position>
+  package var bufferName: String
+  package var macroExpansionEditRange: Range<Position>?
+
+  package init(
+    macroExpansionEditRange: Range<Position>,
+    sourceFileURL: URL,
+    selectionRange: Range<Position>,
+    bufferName: String
+  ) {
+    self.sourceFileURL = sourceFileURL
+    self.selectionRange = selectionRange
+    self.bufferName = bufferName
+    self.macroExpansionEditRange = macroExpansionEditRange
+  }
+
+  private var displayName: String {
+    if let macroExpansionEditRange {
+      "L\(macroExpansionEditRange.lowerBound.line + 1)C\(macroExpansionEditRange.lowerBound.utf16index + 1)-L\(macroExpansionEditRange.upperBound.line + 1)C\(macroExpansionEditRange.upperBound.utf16index + 1).swift"
+    } else {
+      ""
+    }
+  }
+
+  private var queryItems: [URLQueryItem]? {
+    get {
+      let queryItems = [
+        URLQueryItem(name: Parameters.sourceFilePath, value: sourceFileURL.path(percentEncoded: false)),
+        URLQueryItem(name: Parameters.fromLine, value: String(selectionRange.lowerBound.line)),
+        URLQueryItem(name: Parameters.fromColumn, value: String(selectionRange.lowerBound.utf16index)),
+        URLQueryItem(name: Parameters.toLine, value: String(selectionRange.upperBound.line)),
+        URLQueryItem(name: Parameters.toColumn, value: String(selectionRange.upperBound.utf16index)),
+        URLQueryItem(name: Parameters.bufferName, value: bufferName),
+      ]
+
+      return queryItems
+    }
+  }
+
+  package var url: URL {
+    get throws {
+      var components = URLComponents()
+      components.scheme = "sourcekit-lsp"
+      components.host = "swift-macro-expansion"
+      components.path = "/\(displayName)"
+      components.queryItems = queryItems
+
+      guard let url = components.url else {
+        throw ReferenceDocumentURLError(
+          description: "Unable to create URL for macro expansion reference document"
+        )
+      }
+
+      return url
+    }
+  }
+
+  package init(from url: URL) throws {
+    guard url.scheme == "sourcekit-lsp" else {
+      throw ReferenceDocumentURLError(description: "Invalid scheme for reference document")
+    }
```

Comment:
Instead of repeating this scheme check for every reference document URL data that we might get in the future, what do you think about fully splitting the responsibility of the URL components between `ReferenceDocumentURL` and `MacroExpansionReferenceURLData`. 

`ReferenceDocumentURL` is responsible for scheme and host.
`MacroExpansionReferenceURLData` is responsible for display name and query parameters.

So, `MacroExpansionReferenceURLData` would look like the following

```swift
struct MacroExpansionReferenceURLData {
  var displayName: String { … }
  var queryItems: [URLQueryItem] { … }
  
  init(displayName: String, queryItems: [URLQueryItem]) { … }
}
```

And `ReferenceDocumentURL` can do the scheme check + decide which reference document URL type to parse by inspecting the host.

Also, I think we should parse `macroExpansionEditRange` from the display name so that it’s possible to round-trip `ReferenceDocumentURL` to a URL and back to a `ReferenceDocumentURL`.